ros2_foxglove_bridge/src/ros2_foxglove_bridge.cpp

```diff
@@ -1,24 +1,367 @@
+
+        // Remove this topic+datatype tuple
+        _advertisedTopics.erase(topicAndDatatype);
+        _channelToTopicAndDatatype.erase(channel.id);
+
+        // Remove the subscription for this topic, if any
+        _subscriptions.erase(channel.id);
+
+        RCLCPP_DEBUG(this->get_logger(), "Removed channel %d for topic \"%s\" (%s)", channel.id,
+                     channel.topic.c_str(), channel.schemaName.c_str());
+      }
+
+      // Advertise new topics
+      for (const auto& topicAndDatatype : newTopics) {
+        foxglove::ChannelWithoutId newChannel{};
+        newChannel.topic = topicAndDatatype.first;
+        newChannel.schemaName = topicAndDatatype.second;
+
+        try {
+          auto [format, schema] = _messageDefinitionCache.get_full_text(topicAndDatatype.second);
+          switch (format) {
+            case foxglove::MessageDefinitionFormat::MSG:
+              newChannel.encoding = "cdr";
+              newChannel.schema = schema;
+              break;
+            case foxglove::MessageDefinitionFormat::IDL:
+              RCLCPP_WARN(this->get_logger(),
+                          "IDL message definition format cannot be communicated over ws-protocol. "
+                          "Topic \"%s\" (%s) may not decode correctly in clients",
+                          topicAndDatatype.first.c_str(), topicAndDatatype.second.c_str());
+              newChannel.encoding = "cdr";
+              newChannel.schema = schema;
+              break;
+          }
+
+          auto channel = foxglove::Channel{_server.addChannel(newChannel), newChannel};
+          RCLCPP_DEBUG(this->get_logger(), "Advertising channel %d for topic \"%s\" (%s)",
+                       channel.id, channel.topic.c_str(), channel.schemaName.c_str());
+
+          // Add a mapping from the topic+datatype tuple to the channel, and channel ID to the
+          // topic+datatype tuple
+          _advertisedTopics.emplace(topicAndDatatype, std::move(channel));
+          _channelToTopicAndDatatype.emplace(channel.id, topicAndDatatype);
+        } catch (foxglove::DefinitionNotFoundError& err) {
+          RCLCPP_WARN(this->get_logger(), "Could not find definition for topic \"%s\" (%s)",
+                      topicAndDatatype.first.c_str(), topicAndDatatype.second.c_str());
+
+          // Add a mapping from the topic+datatype tuple to a dummy channel so we don't repeatedly
+          // try to load the message definition
+          auto channel = foxglove::Channel{0, newChannel};
+          _advertisedTopics.emplace(topicAndDatatype, std::move(channel));
+        }
+      }
+    }
+
+    if (newTopics.size() > 0) {
+      _server.broadcastChannels();
+    }
+
+    // Schedule the next update
+    _updateCount++;
+    auto nextUpdateMs =
+      std::chrono::milliseconds(std::min(size_t(1) << _updateCount, _maxUpdateMs));
+    _updateTimer = this->create_wall_timer(
+      nextUpdateMs, std::bind(&FoxgloveBridge::updateAdvertisedTopics, this));
```

Comment:
This is truncated exponential backoff, I'll add a comment. I think this is better than a constant period refresh since the bridge is coming online in parallel with the other nodes in the system, so this will result in faster end-to-end iteration speeds when you are frequently restarting the rosgraph.

Polling is annoying since it requires a separate thread and there is a balance between CPU usage and shutdown latency that has to be managed with the `timeout` argument. But it might still be better than this approach. Something to consider for a future PR.

Reply:
https://github.com/foxglove/ros-foxglove-bridge/issues/15